tests/core/test_reader.py

```diff
@@ -94,40 +128,93 @@ def test_sensor_reader(mock_sensor, monkeypatch):
     assert mock_sensor.stubs["sleep"].called
 
 
-def test_sensor_reader_sensor_mismatch(mock_sensor, monkeypatch):
-    sensor_reader = reader.SensorReader(
-        port=mock_sensor.port,
-        samples=0,  # exit immediately
-        sensor="PMSx003",  # match with stubs
-        timeout=0.01,  # low to avoid hanging on failure
+def test_sensor_reader_sleep(sensor_reader_factory, mock_sleep):
+    sensor_reader = sensor_reader_factory(
+        samples=2,  # try to read twice
+        interval=5,  # sleep between samples
     )
 
+    with sensor_reader as r:
+        obs = list(r())
+
+    # check we read twice
+    assert len(obs) == 2
+
+    # check we slept between reads
+    assert 0 < mock_sleep.slept_for < 5
+
+
+def test_sensor_reader_closed(mock_sensor, sensor_reader_factory):
+    sensor_reader = sensor_reader_factory()
+    obs = list(sensor_reader())
+    assert len(obs) == 0
+
+
+def test_sensor_reader_preheat(sensor_reader_factory, mock_sleep):
+    sensor_reader = sensor_reader_factory()
+
+    # override pre heat duration
+    sensor_reader.pre_heat = 5
+
+    with sensor_reader as r:
+        pass
+
+    # check we slept between reads
+    assert mock_sleep.slept_for == 5
+
+
+def test_sensor_reader_warm_up(mock_sensor, sensor_reader_factory, mock_sleep):
+    sensor_reader = sensor_reader_factory()
+
+    def passive_read(n):
+        if n == 1:
+            # first return a "0" payload ("warming up")
+            return (
+                b"BM\x00\x1c"  # expected header
+                + b"\0" * 26  # payload (to total 32 bytes)
+                + b"\x00\xAB"  # checksum
+            )
+        else:
+            # then behave like the original stub again
+            return (
+                b"BM\x00\x1c"  # expected header
+                + b".........................."  # payload
+                + b"\x05W"  # checksum
```

Comment:
I would prefer to move the test setup to a fixture, even if the fixture is only used on this test. This way there is a clear separation between setup and test.